### PR TITLE
Many improvements to docker network and container states

### PR DESCRIPTION
**Pull request by terminalmage**
Much Improved Support for Docker Networking
===========================================

The `docker_network.present` state has undergone a full rewrite, which includes the following improvements:

Full API Support for Network Management
---------------------------------------

The improvements made to input handling in the `docker_container.running` state for 2017.7.0 have now been expanded to `docker_network.present`.  This brings with it full support for all tunable configuration arguments.

Custom Subnets
--------------

Custom subnets can now be configured. Both IPv4 and mixed IPv4/IPv6 networks are supported.

Network Configuration in `docker_container.running` States
-------------------------------------------------------------------

It is now possible to configure static IPv4/IPv6 addresses, as well as links and labels.

Improved Handling of Images from Custom Registries
==================================================

Rather than attempting to parse the tag from the passed image name, Salt will now resolve that tag down to an image ID and use that ID instead.

Due to this change, there are some backward-incompatible changes to image management. See below for a full list of these changes.

Backward-incompatible Changes to Docker Image Management
--------------------------------------------------------

Passing image names to the following functions must now be done using separate
`repository` and `tag` arguments:

- `docker.build`
- `docker.commit`
- `docker.import`
- `docker.load`
- `docker.tag`
- `docker.sls_build`

Additionally, the `tag` argument must now be explicitly passed to the `docker_image.present` state, unless the image is being pulled from a docker registry.

"docker run" Support Added
==================================================

A `docker.run_container` function has been added to mimic `docker run`, running the container and returning the logs from it. In addition, a `docker_container.run` state has been added, which includes many of the same features as the `cmd.run` state (including `onlyif`, `unless`, and `creates` arguments to limit whether or not the container is actually run).

# Tests written?

You betcha! Dozens of new tests, and rewrites of most of the rest due to abstraction of the input translation code.

# Commits signed with GPG?

Yes

### CODE REVIEW

**Comment by terminalmage** [comment]
Note: the integration tests for docker containers and networks are being skipped on Ubuntu 14 because salt-jenkins is not installing docker, docker-py, and busybox like it does for Cent7. Is this something we want to add to salt-jenkins for Ubuntu?

**Comment by rallytime** [comment]
@terminalmage I don't know of any reason off-hand of why we would be limiting those tests to CentOS 7. If you wanted to add that to salt-jenkins, I think that would be just fine.

**Comment by terminalmage** [comment]
https://github.com/saltstack/salt-jenkins/pull/573 adds docker to salt-jenkins for Ubuntu 16.04. I tested this in a vagrant box and all tests passed.

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by terminalmage** [comment]
This has been rebased to include fixes for the unrelated test failures that were merged in the last 24 hours.

**Comment by terminalmage** [comment]
There were some lingering test failures which I believe I have resolved. Tests running again now.

**Comment by rallytime** [comment]
@terminalmage There are 2 docker tests failing on the cent7 build:

- integration.states.test_docker_container.DockerContainerTestCase.test_run_bg
- integration.states.test_docker_network.DockerNetworkTestCase.test_present_attachable

Can you take a look?

File: doc/topics/releases/oxygen.rst [hunk 1]
```diff
@@ -4,6 +4,109 @@
...
+are supported. See :ref:`here <salt-states-docker-network-present-ipam>` for
+more information.
+
+Network Configuration in :py:func:`docker_container.running` States
+-------------------------------------------------------------------
+
+A long-requested feature has finally been added! It is now possible to
+configure static IPv4/IPv6 addresses, as well as links and labels. See
+:ref:`here <salt-states-docker-container-network-management>` for more
+informaion.
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `information`

File: doc/topics/releases/oxygen.rst [hunk 2]
```diff
@@ -4,6 +4,109 @@
...
+State support has also been added via the :py:func:`docker_container.run
+<salt.states.docker_container.run>` state. This state is modeled after the
+:py:func:`cmd.run <salt.states.cmd.run>` state, and includes arguments like
+``onlyif``, ``unless``, and ``creates`` to control whether or not the container
+is run.
+
+Full API Support for :py:func:`docker.logs <salt.modules.dockermod.logs>`
+=========================================================================
+
+This function now supports all of the functions that it's Docker API
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `its`

**Review by cachedout — CHANGES_REQUESTED** [review]

File: doc/topics/releases/oxygen.rst (hunk 1 above)
Comment (terminalmage) [reply doc/topics/releases/oxygen.rst]: /me should have used `:set spell`. 😄 

Thanks!

**Review by rallytime — APPROVED** [review]
There is a lot here, but this is really well done @terminalmage. Nice work.

**Pull request merged**